File: src-tauri/src/agent/native_tools/knowledge_base/list_knowledge_base_documents.rs
//! `list_knowledge_base_documents` 工具（原生）— 列出知识库内全部文档（ID/类型/分块数/状态）。

use crate::agent::native_tools::common::arg_str;
use crate::agent::native_tools::{NativeToolCtx, NativeToolOutcome};
use serde_json::{json, Value};

use super::common::rag_service;

pub(crate) async fn list_knowledge_base_documents_tool(
    _ctx: &NativeToolCtx<'_>,
    args: &Value,
) -> Result<NativeToolOutcome, String> {
    let kb_id = arg_str(args, "knowledge_base_id").unwrap_or_default();
    if kb_id.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: "Missing required parameter: \"knowledge_base_id\". Use list_knowledge_bases to discover available knowledge bases.".to_string(),
            ui_data: None,
        });
    }

    let service = rag_service()?;
    let kb_id_c = kb_id.clone();
    let docs = tokio::task::spawn_blocking(move || service.list_documents(&kb_id_c))
        .await
        .map_err(|e| format!("Task join error: {}", e))?
        .map_err(|e| format!("Error listing documents: {}", e))?;

    if docs.is_empty() {
        return Ok(NativeToolOutcome::Value {
            content: format!("No documents found in knowledge base \"{}\". Use write_to_knowledge_base to create new content, or upload documents through the UI.", kb_id),
            ui_data: Some(json!({ "length": 0, "knowledge_base_id": kb_id })),
        });
    }

    let mut lines = vec![format!("📄 Documents in knowledge base \"{}\" ({} total):", kb_id, docs.len()), String::new()];
    let mut ui_docs: Vec<Value> = Vec::new();

    for (i, d) in docs.iter().enumerate() {
        lines.push(format!("[{}] {}", i + 1, d.file_name));
        lines.push(format!("    Document ID: {}", d.id));
        lines.push(format!("    Type: {}", d.file_type));
        lines.push(format!("    Chunks: {}", d.chunk_count));
        lines.push(format!("    Status: {}", d.status));
        lines.push(String::new());

        ui_docs.push(json!({
            "id": d.id,
            "file_name": d.file_name,
            "file_type": d.file_type,
            "chunk_count": d.chunk_count,
            "status": d.status,
        }));
    }

    lines.push("---".to_string());
    lines.push("Use search_knowledge_base to search within this knowledge base.".to_string());
    lines.push("Use delete_knowledge_base_document with a Document ID to remove it.".to_string());

    Ok(NativeToolOutcome::Value {
        content: lines.join("\n"),
        ui_data: Some(json!({ "length": docs.len(), "knowledge_base_id": kb_id, "documents": ui_docs })),
    })
}

File: src/infrastructure/tools/knowledge-base/list-knowledge-base-documents.ts
/**
 * list_knowledge_base_documents — 列出知识库中的所有文档
 *
 * 返回指定知识库中所有文档的名称、ID、文件类型和片段数。
 * AI 先使用 list_knowledge_bases 发现知识库，
 * 再使用此工具获取文档 ID，然后可用 delete_knowledge_base_document 删除
 * 或用 search_knowledge_base 搜索特定文档内容。
 */
import { toolRegistry } from '@/domain/tools'
import type { ToolExecutor, ToolResult } from '@/domain/tools/types'
import { t } from '@/ui/i18n'
import { ragService } from '@/services/rag-service'

toolRegistry.register(
    'list_knowledge_base_documents',
    (async (args: Record<string, any>, _ctx: any): Promise<ToolResult> => {
    const kbId = args.knowledge_base_id
    if (!kbId || typeof kbId !== 'string') {
      return {
        content:
          'Missing required parameter: "knowledge_base_id". ' +
          'Use list_knowledge_bases to discover available knowledge bases.',
      }
    }

    try {
      const docs = await ragService.listDocuments(kbId)

      if (!docs || docs.length === 0) {
        return {
          content:
            `No documents found in knowledge base "${kbId}". ` +
            'Use write_to_knowledge_base to create new content, ' +
            'or upload documents through the UI.',
          uiData: { length: 0, knowledge_base_id: kbId },
        }
      }

      const lines: string[] = []
      lines.push(`📄 Documents in knowledge base "${kbId}" (${docs.length} total):`)
      lines.push('')

      for (let i = 0; i < docs.length; i++) {
        const d = docs[i]
        lines.push(`[${i + 1}] ${d.file_name}`)
        lines.push(`    Document ID: ${d.id}`)
        lines.push(`    Type: ${d.file_type}`)
        lines.push(`    Chunks: ${d.chunk_count}`)
        lines.push(`    Status: ${d.status}`)
        lines.push('')
      }

      lines.push('---')
      lines.push('Use search_knowledge_base to search within this knowledge base.')
      lines.push('Use delete_knowledge_base_document with a Document ID to remove it.')

      return {
        content: lines.join('\n'),
        uiData: {
          length: docs.length,
          knowledge_base_id: kbId,
          documents: docs.map((d) => ({
            id: d.id,
            file_name: d.file_name,
            file_type: d.file_type,
            chunk_count: d.chunk_count,
            status: d.status,
          })),
        },
      }
    } catch (error: any) {
      return {
        content: `Error listing documents: ${error.message || String(error)}`,
      }
    }
  }) as ToolExecutor,
    t('列出文档'),
)
